scipy/stats/tests/test_stats.py

```diff
@@ -2862,116 +2869,136 @@ def test_zscore_axis(self):
-
-    def test_zscore_2d_all_nan_row(self):
+        xp_assert_equal(z, xp.reshape(stats.zscore(xp.reshape(x, (-1,))), x.shape))
+
+        y = xp.ones((3, 6))
+        with pytest.warns(RuntimeWarning, match="Precision loss occurred..."):
+            z = stats.zscore(y, axis=None)
+        xp_assert_equal(z, xp.full(y.shape, xp.asarray(xp.nan)))
+
+    def test_zscore_constant_input_2d_nan_policy_omit(self, xp):
+        x = xp.asarray([[10.0, 10.0, 10.0, 10.0],
+                        [10.0, 11.0, 12.0, xp.nan],
+                        [10.0, 12.0, xp.nan, 10.0]])
+        s = (3/2)**0.5
+        s2 = 2**0.5
+
+        with pytest.warns(RuntimeWarning, match="Precision loss occurred..."):
+            z0 = stats.zscore(x, nan_policy='omit', axis=0)
+        xp_assert_close(z0, xp.asarray([[xp.nan, -s, -1.0, xp.nan],
+                                        [xp.nan, 0, 1.0, xp.nan],
+                                        [xp.nan, s, xp.nan, xp.nan]]))
+
+        with pytest.warns(RuntimeWarning, match="Precision loss occurred..."):
+            z1 = stats.zscore(x, nan_policy='omit', axis=1)
+        xp_assert_close(z1, xp.asarray([[xp.nan, xp.nan, xp.nan, xp.nan],
+                                        [-s, 0, s, xp.nan],
+                                        [-s2/2, s2, xp.nan, -s2/2]]))
+
+    def test_zscore_2d_all_nan_row(self, xp):
         # A row is all nan, and we use axis=1.
-        x = np.array([[np.nan, np.nan, np.nan, np.nan],
-                      [10.0, 10.0, 12.0, 12.0]])
-        z = stats.zscore(x, nan_policy='omit', axis=1)
-        assert_equal(z, np.array([[np.nan, np.nan, np.nan, np.nan],
-                                  [-1.0, -1.0, 1.0, 1.0]]))
-
-    def test_zscore_2d_all_nan(self):
+        x = xp.asarray([[np.nan, np.nan, np.nan, np.nan],
+                        [10.0, 10.0, 12.0, 12.0]])
+        with pytest.warns(SmallSampleWarning, match="After omitting..."):
+            z = stats.zscore(x, nan_policy='omit', axis=1)
+        xp_assert_close(z, xp.asarray([[np.nan, np.nan, np.nan, np.nan],
+                                       [-1.0, -1.0, 1.0, 1.0]]))
+
+    def test_zscore_2d_all_nan(self, xp):
         # The entire 2d array is nan, and we use axis=None.
-        y = np.full((2, 3), np.nan)
-        z = stats.zscore(y, nan_policy='omit', axis=None)
-        assert_equal(z, y)
+        y = xp.full((2, 3), xp.nan)
+        with pytest.warns(SmallSampleWarning, match="After omitting..."):
+            z = stats.zscore(y, nan_policy='omit', axis=None)
+        xp_assert_equal(z, y)
 
     @pytest.mark.parametrize('x', [np.array([]), np.zeros((3, 0, 5))])
-    def test_zscore_empty_input(self, x):
-        z = stats.zscore(x)
-        assert_equal(z, x)
+    def test_zscore_empty_input(self, x, xp):
+        x = xp.asarray(x)
+        if x.shape[0] == 0:
+            with pytest.warns(SmallSampleWarning, match="One or more sample..."):
+                z = stats.zscore(x)
```

Comment:
All the failures are related to `SmallSampleWarning`s. Here's a bit of information about what's going on there.

Because the function docstring guarantees that it works with array subclasses, I simply switched implementations when `SCIPY_ARRAY_API=1` - the old one handles NumPy arrays and subclasses, the new one is array API compatible.

The new implementation emits `SmallSampleWarning`s sometimes, whereas the old one does not. The tests now look for the warnings, so when `SCIPY_ARRAY_API=1`, they fail. I guess I'll need to special-case the failing tests depending on `SCIPY_ARRAY_API`.

That said, this particular test arguably should not emit `SmallSampleWarning`. The warning message is:
```
/Users/matthaberland/Desktop/scipy/scipy/stats/_stats_py.py:11111: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.
```
because `zscore` calculates the mean and the sample is size zero, but the actual output is empty (no NaNs), so there probably shouldn't be a warning.

Also, upon further reflection, there probably shouldn't be a `SmallSampleWarning` in the test above either. `nan_policy='omit'` only makes sense in the reduction operations (mean and std) that happen internally, but since this is an elementwise function, it should not be surprising that NaNs in the input produce NaNs in the output.